x/hold/hold_test.go

```diff
@@ -0,0 +1,70 @@
+package hold
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/testutil/assertions"
+)
+
+func TestAccountHold_Validate(t *testing.T) {
+	cz := func(coins string) sdk.Coins {
+		rv, err := sdk.ParseCoinsNormalized(coins)
+		require.NoError(t, err, "ParseCoinsNormalized(%q)", coins)
+		return rv
+	}
+	c := func(amount int64, denom string) sdk.Coin {
+		return sdk.Coin{Denom: denom, Amount: sdkmath.NewInt(amount)}
+	}
+	ah := func(addr string, amount sdk.Coins) AccountHold {
```

Comment:
You could probably make a constructor for this to make it a bit more readable.

Reply:
I agree here.  I'm don't really like this pattern of using an anonymous function as a constructor instead of
 using production code constructors in tests.  I can easily see mistakes being done creating these and they are a bit harder to read.
Also, the intent of using constructors was to provide a single path to initialize objects into the proper state for production side code.  Therefore, when the constructor changes due to new fields being added, all usages of the old constructor will break.  I find a lot of value in this.

Reply:
I didn't make a `NewAccountHold(addr, coins)` because for production code, I'd need the `addr` to be an `sdk.AccAddress`. In these unit tests, though, I need to be able to directly set the string so that I can test when it's not valid. So even if I had made `NewAccountHold`, I wouldn't be helpful in here. And in the production code, it was handier to just use struct initialization to make them.

There's four primary reasons why I use anonymous funcs in the tests like this:
1. Smaller diff when a field is added. **If I didn't break a test case, I shouldn't have to make a change to it.** When a field is added and the production constructor gets a new argument, you have to update all uses of that constructor. If you have to update all tests cases just so it'll compile, it makes it harder to notice (in the diff) which test cases actually broke and needed a logical change. Identifying those is an important part of evaluating the correctness a change. Plus, a smaller diff means that you could make your change faster, and requires less time to review.
2. Provide a constant to one or more fields. If every test case has the same value for a certain thing, it's better to move that value out of the test case definition. That way, the next person to look at the test case doesn't have to look at the value and wonder what bearing it has on that specific test case. If it's inconsequential to the case, it shouldn't be part of the case definition.
3. Error checking. The anonymous funcs can call `require.NoError` and return only the thing we want. This is preferable to using funcs that might panic because panics can disrupt an entire chunk of tests (not just the one `Test...` func or test case the panic occurs in). By removing the `error` from the return values, it makes it easier to define things specific to a test case in the test case. I.e. you don't have to create it above, handle the error, then use the variable in the test case.
4. Brevity. Test cases can be pretty complex. Removing commonly repeated, easy to deduce text can help make a test case easier to understand; it can improve the signal-to-noise ratio. Being able to identify the differences between two test cases is made easier when it's easier to see both at the same time, and when you have less to ignore. This identification is most important when a test case unexpectedly fails, and you need to know what makes it special.

Honestly, 95% of our constructors are worthless. In some cases, they actually make things worse. The only ones that are worthwhile are ones that either set private fields or ones that do some conversion or processing (other than calling `.String()`). If there's more than a few fields, they actually make things worse because you can't call functions using named parameters, so you're having to count arguments to see which argument in the call gets set to which field. This is made even worse when there's multiple arguments of the same type in a row.

Most of our stuff doesn't have anything private or require any extra processing, though. As far as I can tell, the only reason we have most of those constructors is because a lot of us came from the java world, where you don't have struct initialization like you do in go. And since go doesn't have function overloading, you're limited to a single `New...` constructor, which quite often ends up not being useful; we end up jumping through hoops to be able to use it when we could just be using struct initialization and save ourselves some effort.

In production code, brevity isn't nearly the concern it is in test cases. It's usually better to have named parameters that you get with struct initialization. Test cases are like a table of data where it's useful to be able to see several similar things at once. In the production code, it's usually not as important to compare the differences between two lines. In the cases where that is important, usually the verboseness of struct initialization with named arguments is more helpful than trying to compare function arguments, specially if the lines are far apart or when there's so many arguments, they're spread across multiple lines.

In unit tests, it's best to isolate the thing your testing as much as possible. When you use a production constructor in a test, now you're test relies on that constructor too. If there's a problem with the constructor, you only want the constructor tests to fail. You don't want other unrelated tests to also fail because it muddies the picture of what's actually been broken. While it's sometimes useful to have a case or two covering that interaction (e.g. a test case on `DefaultGenesis().Validate()`), most of the test cases should not have such a dependency.

Ultimately, the production constructors cause problems, extra work, and little-to-no benefit.

Reply:
I did go through and rename the `ah` funcs to `accountHold` as well as the `c` and `cz` funcs.

Reply:
My main concern was the readability of it because I was having a hard time following the tests, and I was seeing tests redefine the same anonymous functions. A constructor was just one proposed solution., but it seems like you fixed my concern. 

I find benefits in both approaches, but constructors do have some additional benefits.

1. You can ensure object creation all follows the same type of creational pattern. For example all types must follow this, all non pods must follow this, no types must follow this, etc...
2. Removal of code redundancy with New and Init functions.
3. This goes against your first point, and you cannot always guarantee a test is valid unless the objects that it uses are in a correct state. It's very easy to add a new mandatory field to a struct, and forget to update all instances of the manually initialized struct. The compiler cannot guide the user in this case because there is no constructor, and it is up to the user to update their tests to ensure all tests are in valid starting state. These passing tests may or may not be working anymore. See the following example:

```
type S struct {
	a string
	b string // Newly added mandatory field
}

func Foo(s S) int {
	// Oops, I acidentally commented this out
	/*if len(s.a) == 0 {
		return 1
	}*/ 

	// Newly added code
	if len(s.b) == 0 {
		return 1
	}

	return 0
}

// This test is now broken without the compiler telling you
// It can be very hard to find all instances of it in bigger projects
// We can easily pass an S through this without a valid value for A.
func TestAFooIsNotInvalid(t *testing.T) {
	s := S {}
	res := Foo(s)

	// This passes, but the code is not correct.
	// An invalid A will make it past the check
	assert(res, 1)
}
```

It might be better to discuss this offline as a team so that if we are choosing to go constructors or no constructors in tests, then we know all the benefits/risks we are getting from this code style.